jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/DataUtils.scala

```diff
@@ -73,12 +74,81 @@ object DataUtils extends Serializable {
     }
   }
 
+  private def featureValueOfDenseVector(rowHashCode: Int, features: DenseVector): Float = {
+    val featureId = {
+      if (rowHashCode > 0) {
+        rowHashCode % features.size
+      } else {
+        // prevent overflow
+        math.abs(rowHashCode + 1) % features.size
+      }
+    }
+    features.values(featureId).toFloat
+  }
+
+  private def featureValueOfSparseVector(rowHashCode: Int, features: SparseVector): Float = {
+    val featureId = {
+      if (rowHashCode > 0) {
+        rowHashCode % features.indices.length
+      } else {
+        // prevent overflow
+        math.abs(rowHashCode + 1) % features.indices.length
+      }
+    }
+    features.values(featureId).toFloat
+  }
+
+  private def calculatePartitionKey(row: Row, numPartitions: Int): Int = {
+    val Row(_, features: Vector, _, _) = row
+    val rowHashCode = row.hashCode()
+    val featureValue = features match {
+      case denseVector: DenseVector =>
+        featureValueOfDenseVector(rowHashCode, denseVector)
+      case sparseVector: SparseVector =>
+        featureValueOfSparseVector(rowHashCode, sparseVector)
+    }
+    math.abs((rowHashCode.toLong + featureValue).toString.hashCode % numPartitions)
```

Comment:
@trams please check this 

this has been `Long + Double and toString and hashCode `

it shall prevent overflow

Reply:
Long + Double is a double, mod a double with an integer and then to int creates many conflicts in partition number

e.g. any value in [1.0, 2.0) will be assigned to partition 1....

(actually rowHashCode.toDouble + featureValue is more helpful for reading

Reply:
I see your point but I do not think that any feature value with `[1.0, 2.0)` will be assigned a partition 1 given that this featureValue should have different hash codes

I do not mind toString if it is better but I would prefer to do without it.

What about using `rowHashCode.toLong + featureValue.toLong`

P.S. I really do not have an experience building hash functions and I am always paranoid to accidentally take the bad one so I am fine with toString. I just want to make sure we looked at all options

Reply:
Also what do you think about completely different option: "Use murmur to combine rowHashCode and featureValue into a good hash value"

P.S. I am just spitballing here so the idea may be completely stupid

Reply:
AFK, just quick response to why [1,2) will be assigned to the same partition

Your propsoal is 
math.abs((rowHashCode.toLong + featureValue) % numPartitions.toLong).toInt

Because of toInt here, 1.1.toInt will be 1, 1.6.toInt will be the same

Reply:
I understand that 1.1.toInt = 1.6.toInt = 1 but it is not obvious to be it is a problem to apply `.toInt` to `featureValue`. Yes in case
1.rowHashCode.toLong = x, featureValue = 1.1 we would have featureValue.toInt = 1 the result will be (x + 1) % numPartitions
2. .rowHashCode.toLong = y, featureValue = 1.6 we would have featureValue.toInt = 1 the result will be (y + 1) % numPartitions

given that x != y because featureValue is part of Row and different it should (key word should) not be a problem

But I am pretty sure I am missing something else here. I do not want to bikeshed here and prematurely optimize (my concern with toString is mostly about performance and excessive allocations of Strings) cause I think this change is awesome and will solve a lot of problems

All this discussion made me think. Do we have any tests on "real" data which shows that this hash function is "uniform" to some extend?